ChangeDependencyGroupIdAndArtifactId generates empty diffs on Java files after #7202

## What happened

- After merging #7202 (Expand star imports in ChangePackage), `ChangeDependencyGroupIdAndArtifactId` now updates `JavaSourceSet` markers on Java source files in its `getVisitor()` phase. This causes the test framework to detect "empty diff" errors in downstream repos where tests don't provide explicit `after` text for Java source files.

The error: `AssertionError: An empty diff was generated. The recipe incorrectly changed a reference without changing its contents.`

## Affected downstream repos

All started failing on April 11, 2026 (were green on April 10):

- **openrewrite/rewrite-hibernate** — `MigrateToHibernate61Test.groupIdHibernateOrmRenamed()`
- **openrewrite/rewrite-migrate-java** — `AddCommonAnnotationsDependenciesTest.changeAndUpgradeDependencyIfAnnotationJsr250Present()`
- **moderneinc/rewrite-dropwizard** — `MigrateDropwizardToSpringBoot3Test.addsSpringStarterJdbcWhenDropwizardDbIsUsed()`

## Root cause

`ChangeDependencyGroupIdAndArtifactId.getVisitor()` now processes Java files (line 350-351) and calls `updateJavaSourceSet()`, which modifies the `JavaSourceSet` marker via `JavaSourceSet.updateOnSourceFile()`. This returns a new `SourceFile` object (different identity), causing the recipe framework to generate a `Result`. However, the printed text is unchanged — only markers differ — so `RewriteTest` flags it as an "empty diff" error.

The PR's own test (`updatesJavaSourceSetMarkerOnJavaFiles`) handles this by providing explicit `after` text identical to `before`. But downstream tests don't know about this new behavior.

## Suggested fix

Consider one of:

1. **Filter marker-only Results in `InMemoryLargeSourceSet.getChangeset()`** — use `Result.isLocalAndHasNoChanges()` to skip Results where text is unchanged. This would need adjustment to the `updatesJavaSourceSetMarkerOnJavaFiles` test (use `afterRecipe` validation instead of `after` text matching).

2. **Relax the "empty diff" check in `RewriteTest`** — instead of failing when a Result has no text change and no `after` spec, silently ignore the Result. This is a less invasive change but reduces bug detection coverage.

3. **Move marker updates out of the visitor phase** — apply JavaSourceSet updates through a mechanism that doesn't generate Results (e.g., execution context sharing, or a post-processing step).

## Workaround

Downstream repos can add explicit `after` text (identical to `before`) for Java source files in affected tests. PRs are being submitted for the affected repos.

--- a/rewrite-java/src/main/java/org/openrewrite/java/marker/JavaSourceSet.java
+++ b/rewrite-java/src/main/java/org/openrewrite/java/marker/JavaSourceSet.java
@@ -66,6 +66,11 @@ public class JavaSourceSet implements SourceSet {
      * @return a new JavaSourceSet with the types added
      */
     public JavaSourceSet addTypesForGav(String gavKey, List<JavaType.FullyQualified> types) {
+        List<JavaType.FullyQualified> existing = gavToTypes.get(gavKey);
+        if (existing != null && existing.equals(types)) {
+            return this;
+        }
+
         List<JavaType.FullyQualified> newClasspath = new ArrayList<>(classpath);
         newClasspath.addAll(types);
 
@@ -178,6 +183,9 @@ public class JavaSourceSet implements SourceSet {
         String cacheKey = maybeJp.get().getId().toString() + ":" + maybeSourceSet.get().getName();
         JavaSourceSet cached = cache.get(cacheKey);
         if (cached != null) {
+            if (cached == maybeSourceSet.get()) {
+                return sf;
+            }
             return sf.withMarkers(sf.getMarkers().setByType(cached));
         }
         JavaSourceSet updated = transform.apply(maybeSourceSet.get());

--- a/rewrite-test/src/main/java/org/openrewrite/test/RewriteTest.java
+++ b/rewrite-test/src/main/java/org/openrewrite/test/RewriteTest.java
@@ -562,7 +562,15 @@ public interface RewriteTest extends SourceSpecs {
                             boolean isRemote = result.getAfter() instanceof Remote;
                             if (!isRemote && Objects.equals(result.getBefore().getSourcePath(), result.getAfter().getSourcePath()) &&
                                 Objects.equals(before, actualAfter)) {
-                                fail("An empty diff was generated. The recipe incorrectly changed a reference without changing its contents.");
+                                // Marker-only change (e.g. updated JavaSourceSet classpath) — no visible diff.
+                                // Still call afterRecipe so marker assertions work, then skip.
+                                allResults.remove(result);
+                                try {
+                                    //noinspection unchecked
+                                    ((Consumer<SourceFile>) sourceSpec.afterRecipe).accept(result.getAfter());
+                                } catch (ClassCastException ignored) {
+                                }
+                                continue nextSourceFile;
                             }
 
                             assert result.getBefore() != null;

